patches/tModLoader/Terraria/ModLoader/ArmorSet.cs

```diff
@@ -0,0 +1,105 @@
+﻿using Microsoft.Xna.Framework;
+using System;
+using System.Collections.Generic;
+using Terraria;
+using Terraria.ModLoader;
+
+namespace Terraria.ModLoader
+{
+	public abstract class ArmorSet : ModType
+	{
+		/// <summary>
+		/// The numerical IDs of the equipment needed for the set bonus to take effect.<br/>
+		/// Can store the IDs for a head piece, a body piece, and a leg piece.<br/>
+		/// </summary>
+		public abstract (int? head, int? body, int? legs) RequiredEquipment { get; }
+
+		/// <summary>
+		/// The text displayed to describe what this set bonus does.<br/>
+		/// </summary>
+		/// <param name="color">
+		/// Allows the specification of a color in which to draw the description for this set bonus.<br/>
+		/// </param>
+		public abstract string SetBonusDescription(ref Color? color);
+
+		/// <summary>
+		/// The method used to apply FUNCTIONAL set effects.<br/>
+		/// </summary>
+		/// <param name="player">
+		/// The player to apply these functional set effects to.<br/>
+		/// </param>
+		public abstract void ApplyFunctionalEffects(Player player);
+
+		/// <summary>
+		/// The method used to apply VANITY set effects.<br/>
+		/// </summary>
+		/// <param name="player">
+		/// The player to apply these vanity set effects to.<br/>
+		/// </param>
+		public abstract void ApplyVanityEffects(Player player);
+
+		/// <summary>
+		/// Checks whether or not the player has the given armor set equipped in functional slots.<br/>
+		/// By default, checks for all pieces included in the set; if at least one is missing, the functional effects of this set are not applied.<br/>
+		/// Override this only if you need to make your own active check for this armor set's functional effects.<br/>
+		/// </summary>
+		/// <param name="player">
+		/// The player to check set bonus validity for.<br/>
+		/// </param>
+		/// <returns>
+		/// If the conditions to apply the functional armor set bonus are met, <see langword="true"/>; otherwise, <see langword="false"/>.</returns>
+		public virtual bool ActiveFunctional(Player player) {
```

Comment:
I wouldn't put this method here, and I definitely wouldn't make it `virtual`.

Instead, if we're going for this, it should be as simple as `ArmorSetLoader.FindMatchingSet(int head, int body, int legs)`

`FindMatchingSet` _can_ loop through all of the armor sets, and compare this, but we could also organise armor-sets into a tree and make the lookup much much faster.